Feat(docs):The Entity Editor overview and Adding an Entity workflow (#3)
metabrainz/bookbrainz-dev-docs: docs/entity-editor.rst

######################
The Entity Editor
######################

Introduction
==============

In Bookbrainz, a set of information about a particular author, book, publication, magazine, etc. is referred to as an *Entity*, and the form which we use to add/edit this information is called the **Entity-Editor**.
There are 6 entity types in the BookBrainz world: Author, Work, Edition Group, Edition, Publisher and Series.
For a better description of each entity, see :ref:`entities-description`.

> The code for the entity editor can be found in ``src/client/entity-editor``.

Reviewer: I think it would be useful to make this a link to the folder in question:

```suggestion
The code for the entity editor can be found in `src/client/entity-editor <https://github.com/metabrainz/bookbrainz-site/tree/master/src/client/entity-editor>`_.
```

The entity-editor has been divided into different sections, and each of these sections have a folder within this directory.

> We use Redux for state management. Each section has its own set of actions and reducers to handle state corressponding to the component. For more information on how Redux works, you can refer to this: `Getting started with React-Redux <https://react-redux.js.org/introduction/getting-started>`_.

Reviewer: ```suggestion
We use Redux for state management. Each section has its own set of actions and reducers to handle state corresponding to the component. For more information on how Redux works, you can refer to this: `Getting started with React-Redux <https://react-redux.js.org/introduction/getting-started>`_.
```

> The directory for each section contains the separate components(which share common code) for editing and merging entities, along with the aforementioned actions and reducers. The validation code is also present in the folder.

Reviewer: Some suggestions to really spell it out: 
```suggestion
Each section has its own directory containing the React components for editing and merging entities, along with the aforementioned Redux actions and reducers. You will also find a ``common`` folder with shared code 
 and a ``validators`` folder containing the code used to validate the data entered in the form before submission.
```

Different sections of the Entity Editor
========================================

The Entity-Editor can be broken down into distinct mini-sections which are as follows:

Name Section
-------------
Alias Section
--------------
Identifier Section
----------------------
Entity Section
------------------
Relationship Section
---------------------
Annotation Section
---------------------
Submission Section
----------------------

Adding an Entity
=================

When we click on an Add Entity button, it takes us to a page with an url like this:
``https://bookbrainz.org/{entityType}/create``.

> We make use of this ``entityType`` to generate the entity-specific markup for our entity-editor.

Reviewer: I would also add something along the lines of:

```
The `/create` routes are created separately for each entity type, each having a file in `the src/server/routes/entity folder <https://github.com/metabrainz/bookbrainz-site/tree/master/src/server/routes/entity>`_.
```

While creating the entity specific markup for our entity-editor, we also inject certain props into our entity-editor. 

> We make use of middlewares to load ``languageTypes``, ``identifierTypes``, and ``relationshipTypes`` specific to the entity we are going to add. 

Reviewer: ```suggestion
We make use of `router-level middleware <https://expressjs.com/en/guide/using-middleware.html#middleware.router>`_ to load ``languageTypes``, ``identifierTypes``, and ``relationshipTypes`` specific to the entity we are going to add. 
```

We might also need to load some entity-specific props, for example, in case of Work entity, we load ``workTypes`` too.

> While creating the markup, it also creates a store for our entity-editor, with a *rootReducer* which consists of a combination of reducers, each concerned with a particular section of the entity-editor namely:

Reviewer: ```suggestion
While creating the markup, it also creates a Redux store for our entity-editor, with a *rootReducer* which consists of a combination of reducers, each concerned with a particular section of the entity-editor namely:
```

* aliasEditorReducer
* annotationSectionReducer
* buttonBarReducer
* entityReducer
* identifierEditorReducer
* nameSectionReducer
* relationshipSectionReducer
* submissionSectionReducer

Let us take an example how we make use of actions and reducers in our entity-editor by describing the workflow for the **Name-Section** of the entity editor.

You can take a look at the initial state of a particular section by looking at the reducer in its directory. 

> The initial state for Name-section can be seen by looking at the reducer file in ``src/client/entity-editor/name-section/reducer.js``. The initial state looks like this: 

Reviewer: Maybe make the path a link to the file here too?

* ``onNameChange``: this updates the *nameValue*.
* ``onNameChangeSearchName``: as we enter the name, we try to seach the nameValue in our database in order to let the user see whether the entity already exists in our database. This updates the search term and adds the results in *searchResults*.
* ``onNameChangeCheckIfExists``: if we find an entity whose name matches exactly with the entity name we entered, we try to display a warning so as to avoid the user from making a duplicate entry. This updates the content of *exactMatches*.
* ``searchForMatchindEditionGroups``: Search for Edition Groups that match the name, if the entity is an Edition.

Similarly, appropriate eventHandlers and actions are present for updating the value of *Sort Name* field, *Language*, and *Disambiguation*.

Hence, a similar pattern is followed on all the other sections of the entity-editor, where we make use of onChange event handlers for a particular field, to fire off an action with the help of the dispatch function to update its corresponding field in the state.

    
The Relationship Editor
-------------------------

> Any type of connection between two entities in BookBrainz is called a relationship. We make use of a Modal to add relationships to the entity we are adding. 

Reviewer: Here I think  a link to the component is more useful than describing how it is presented considering we describe it below.

```suggestion
Any type of connection between two entities in BookBrainz is called a relationship. We use the `relationship editor component <https://github.com/metabrainz/bookbrainz-site/tree/master/src/client/entity-editor/relationship-editor>`_ to add relationships to the entity we are creating.
```

* Rendering a list of already added relationships. This is done with the help of *RelationshipList* component present in ``src/client/entity-editor/relationship-editor/relationship-section.tsx``

We make use of ``relationshipEditorVisible`` flag to toggle the Relationship editor modal. Within the relationship editor modal, there are two fields:

**Entity Select field** : The *renderEntitySelect* function deals with this field. Here ``baseEntity`` is the entity which is being edited. The ``EntitySearchFieldOption`` allows us to search for any existing entity which we would like to link to our current *baseEntity*.

We can apply some additional filters to our search, so as to the optimize search results. For example, in case of a Series entity of type X, we dont need to display search results with entities which are not of the same time.
When we select an entity from the Search results, it gets stored as ``targetEntity``.

**RelationshipType Select field** : After selecting a targetEntity, we make use of a function called ``generateRelationshipSelection`` which takes our relationshipTypes object which was passed as a prop to our entity-editor, the baseEntity, and the targetEntity.

> This function then tries to generate all possible relationships which between the two entities. We can then select the Relationship Type for our entity using the RelationshipSelect field in the editor. This sets the value of ``relationship`` and ``relationsipType`` of our state.

Reviewer: ```suggestion
This function returns all combinations of relationship types which are valid between the two entities. We can then select the Relationship Type for our entity using the RelationshipSelect field in the editor. This sets the value of ``relationship`` and ``relationsipType`` of our state.
```

When we click on Add, we pass the ``relationship`` object to the following action: 
::
    let nextRowID = 0;
    export function addRelationship(data: Relationship): Action {
        return {
            payload: {data, rowID: `n${nextRowID++}`},
            type: ADD_RELATIONSHIP
        };
    }

> Here the data is the ``relationship`` object we passed from our Relationship Editor props. In the payload for this action, we also pass a ``rowID``.

Reviewer: ```suggestion
Here the data is the ``relationship`` object we passed from our Relationship Editor props. In the payload for this action, we also pass a ``rowID`` which is the index of the relationship in the array of relationships for that entity.
```

This is then added to the ``relationships`` object, with the ``rowId`` acting as a key for the mapping.

As we keep on adding relationships, they are rendered as a list on the entity-editor with the help of the aforementioned RelationshipList component. We can still edit and remove these relationships from the list.
If we click the edit button next to a particular relationship, it opens up the relationship modal with the ``relationship`` object passed as prop to the relationshipEditor.

Submission Section
---------------------

> When we click on submit, the entire state of the form(rootState) is sent to an appropriate *createHandler*(``{entityType}/create/handler``). If all goes well, we are redirected to the Display Page of our current entity.

Reviewer: I guess more details here wouldn't hurt :p

It's worth pointing out that we move from front-end code to server-side code

Quick outline:
- a handler function is created with the `makeEntityCreateOrEditHandler` utility function (link to src/server/helpers/entityRouteUtils.tsx ?) adapted to the entity type
- the rootState is validated using the validators for that entity type
- the rootState is manipulated to be in a certain format by the `transformNewForm` function in each entity's route definitions file (`src/server/routes/entity/{entityType}`)
-  that formatted form is passed to the `handleCreateOrEditEntity` function which takes care of creating and modifying the required ORM models [<link to docs/ORM>] (including other entities i.e. for relationships) and persisting the changes to the database.
- If all the above steps are successful, the user is redirected to the Display Page of the newly created entity.

######################
The Entity Editor
######################

Introduction
==============

In Bookbrainz, a set of information about a particular author, book, publication, magazine, etc. is referred to as an *Entity*, and the form which we use to add/edit this information is called the **Entity-Editor**.
We currently have the following types of entities in Bookbrainz:

* Author
* Work
* Edition
* Edition-Group
* Series 

> * Publisher

Reviewer: I would use this (which has a link to the page where we describe more) which I copied from another page:
```
There are 6 entity types in the BookBrainz world: Author, Work, Edition Group, Edition, Publisher and Series.
For a better description of each entity see :ref:`entities-description`.
```

Author: Makes sense. I have made the changes, as suggested.

######################
The Entity Editor
######################

Introduction
==============

In Bookbrainz, a set of information about a particular author, book, publication, magazine, etc. is referred to as an *Entity*, and the form which we use to add/edit this information is called the **Entity-Editor**.
We currently have the following types of entities in Bookbrainz:

* Author
* Work
* Edition
* Edition-Group
* Series 
* Publisher

> 

Reviewer: It would be good here to give a broad overview of the editor without going into details:
* the code for the editor is in its own folder in the codebase
* we use Redux for state management
* validation code is also in folder
* there are separate components (but that share common code) for editing entities and for merging entities
* Anything else?

Author: That seems to cover it all. Have added this.

Different sections of the Entity Editor
========================================

The Entity-Editor can be broken down into distinct mini-sections which are as follows:

Name Section
-------------
Alias Section
--------------
Identifier Section
----------------------
Entity Section
------------------
Relationship Section
---------------------
Annotation Section
---------------------
Submission Section

> ----------------------

Reviewer: :+1:

If these are going to be filled up afterwards, we can create a table of contents above all these:
```
* :ref:`Name Section`
* :ref:`Alias Section`
etc.
```

This will create a list where each item is a link to the appropriate section

Author: Yeah. I was thinking of adding some information about what fields/variables each section deals with. Then we can provide a link to some other part of the documentation where we explain what that particular section is all about (without going into the coding part). Something like `editing.rst` would be a good place for that, I think. There, we can provide information about how to make good and useful edits on Bookbrainz. This would involve explaining all the different sections of the editor.

Reviewer: So we have separate documentation for how to use the website and enter edits: https://bookbrainz-user-guide.readthedocs.io/en/latest/
It's pretty empty for the moment being, but that's what you describe: user documentation as opposed to developer documentation

Reviewer: I'm still not clear on the formatting here. Are they meant to be sub-headers?

If not it could be formatted as a definition list instead by indenting a small blurb of text under each list item:
```
Name Section
  Used to add a required default alias for the entity
Alias Section
  Used to enter additional aliases such as pseudonyms, other languages and scripts, etc.
```

Which will be rendered like this:
![image](https://user-images.githubusercontent.com/6179856/151165283-d96be685-8252-47bf-a2e0-99fb60b1b94a.png)

Author: > 

I have not completed this section yet. Initially, I wanted to give an outline of what fields a particular section covers, because we are not describing each and every section in the workflow that follows. So yeah, something like this.
